File: 2020/day25/main.py
def main():
    print(solve_1())

MODULUS = 20201227

def solve_1():
    pub_a = 8252394
    pub_b = 6269621

    a = bruteforce(pub_a)
    b = bruteforce(pub_b)

    return (a, b)

    # print(transform(7, 8))
    # print(modular_pow(7, 8, MODULUS))





    # enc_squared = (pub_a * pub_b) % MODULUS

    # enc = modular_pow(enc_squared, (MODULUS + 1) // 4, MODULUS)
    # return enc


MAX_BRUTE = 100_000_000

def bruteforce(public):
    print("Bruteforcing ", public)
    for i in range(MAX_BRUTE):
        if i % (MAX_BRUTE // 100) == 0:
            print(i)
        if modular_pow(7, i, MODULUS) == public:
            return i
    
    return -1

def transform(subject_number, loop_size):
    value = 1
    for _ in range(loop_size):
        value *= subject_number
        value %= 20201227
    return value

def modular_pow(base, exponent, modulus):
    if modulus == 1:
        return 0
    c = 1

    for e_prime in range(0, exponent):
        c = (c * base) % modulus

    return c

def handshake(card_loop_size, door_loop_size):
    card_public = transform(7, card_loop_size)
    door_public = transform(7, door_loop_size)


if __name__ == "__main__":
    main()

File: 2020/day25/main.c
#include <stdio.h>

#define MODULUS 20201227

typedef unsigned long long u64;


u64 transform(u64 subject_number, u64 loop_size) {
    u64 i;
    u64 value = 1;

    for (i = 0; i < loop_size; i++) {
        value *= subject_number;
        value %= MODULUS;
    }

    return value;
}

u64 bruteforce(u64 target) {
    u64 loops = 0;
    u64 value = 1;

    printf("Value: %d\n", value);
    printf("Target: %d\n", target);
    while (value != target) {
        value *= 7;
        value %= MODULUS;
        loops++;
    }

    return loops;
}


int main() {
    u64 pub_a = 8252394;
    u64 pub_b = 6269621;
    printf("Bruteforcing...\n");
    u64 sec_a = bruteforce(pub_a);

    printf("sec_a: %llu\n");
    printf("enc: %llu\n", transform(pub_b, sec_a));

    return 0;
}